lib/app/dispatcher.rb

```diff
@@ -0,0 +1,88 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License;
+# you may not use this file except in compliance with the Elastic License.
+#
+
+# frozen_string_literal: true
+
+require 'active_support'
+require 'concurrent-ruby'
+require 'connectors'
+require 'core'
+require 'utility'
+require 'app/config'
+
+module App
+  class Dispatcher
+    POLL_IDLING = (App::Config[:idle_timeout] || 60).to_i
+    TERMINATION_TIMEOUT = (App::Config[:termination_timeout] || 60).to_i
+
+    attr_reader :workers
+
+    def initialize
+      @workers = {}.with_indifferent_access
+      @pool = Concurrent::ThreadPoolExecutor.new(
+        min_threads: 5,
+        max_threads: 10,
+        max_queue: 100,
+        fallback_policy: :caller_runs
+      )
+      @is_shutting_down = false
+    end
+
+    def start!
+      if App::Config['mode'] == 'dispatcher'
+        loop do
+          connectors = Core::ElasticConnectorActions.native_connectors
+          if connectors.empty?
+            Utility::Logger.info('No native connectors found.')
+            return
+          end
+          Utility::Logger.info("Total workers: #{workers.size}. Checking if all native connectors are running...")
+          connectors.each do |connector|
+            if !workers[connector[:id]].nil?
```

Comment:
How do you reset the worker?

Reply:
What does it mean to reset the worker?

Reply:
https://github.com/elastic/connectors-ruby/blob/a841b92fd6b0ec8a16da4d6524f780aa379b64a8/lib/app/dispatcher.rb#L43-L44

With the above code, I suppose what you want to avoid is to have 2 jobs running for the same connector. But you must have a way to reset `@workers` when the job completes, so that `@workers[connector[:id]]` is `nil`. Otherwise there won't be any job running for a particular connector after the first job.

I don't think we need this check here. Even though there's no active job running for a connector in this connector service, there might be another one running in another instance of connector service, and we should make the connectors resilient to multi connector services (not a goal for 8.5)

Reply:
I don't quite get this.

> But you must have a way to reset `@workers` when the job completes, so that `@workers[connector[:id]]` is nil. Otherwise there won't be any job running for a particular connector after the first job.

To me, it doesn't sound true, because the worker never stops. It keeps polling for jobs until the program is killed or fails with an unrecoverable exception, AFAIK. A job might be over, but the worker will still keep polling for a new one.

The note you made brings up another concern though, and that is, what if the connector/source is removed? Then it should go away, as in it won't be present in the `.elastic-connectors`, and __then__ the dispatcher should nullify that worker (and kill the thread). For that, I don't provision now, and I suppose I should.